src/python/tests/assembly/test_download.py

```diff
@@ -0,0 +1,458 @@
+import filecmp
+import logging
+from pathlib import Path
+from typing import Callable, ContextManager, Optional
+from unittest.mock import Mock, patch, MagicMock
+
+from ftplib import error_reply as ftp_error_reply
+import pytest
+
+from ensembl.io.genomio.assembly.download import UnsupportedFormatError
+from ensembl.io.genomio.assembly.download import FileDownloadError
+from ensembl.io.genomio.assembly.download import FTPConnectionError
+from ensembl.io.genomio.assembly.download import establish_ftp
+from ensembl.io.genomio.assembly.download import get_checksums, md5_files, _download_file
+from ensembl.io.genomio.assembly.download import download_files, get_files_selection, retrieve_assembly_data
+
+
+#################
+@pytest.mark.dependency(name="test_ftp_connection")
+@pytest.mark.parametrize(
+    "ftp_url, accession, expectation",
+    [
+        pytest.param(
+            "ftp.ncbi.nlm.nih.gov", "GCA_017607445.1", does_not_raise(), id="Successful ftp connection"
+        ),
+        pytest.param(
+            "", "GCA_017607445.1", pytest.raises(FTPConnectionError), id="Failed connection case bad url"
+        ),
+        pytest.param(
+            "ftp.ncbi.nlm.nih.gov",
+            "gcx_017607445.1",
+            pytest.raises(UnsupportedFormatError),
+            id="Failed connection improper INSDC accession",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
+def test_ftp_connection(
+    mock_ftp: Mock,
+    ftp_url: str,
+    accession: str,
+    expectation: ContextManager,
+):
+    """Tests the FTPConnection method `establish_ftp()`.
+
+    Args:
+        mock_ftp: Mock FTP object.
+        ftp_url: FTP URL.
+        sub_dir: Subdirectory path.
+        expectation: Context manager expected raise exception.
+    """
+
+    def side_eff_conn(url: str):
+        if not url:
+            raise FTPConnectionError()
+
+    mock_ftp.connect.side_effect = side_eff_conn
+
+    with expectation:
+        connected_ftp = establish_ftp(mock_ftp, ftp_url, accession)
+        connected_ftp.connect.assert_called_once_with(ftp_url)  # type: ignore[attr-defined]
+        connected_ftp.login.assert_called_once()  # type: ignore[attr-defined]
+        connected_ftp.cwd.assert_called_once_with("genomes/all/GCA/017/607/445")  # type: ignore[attr-defined]
+
+
+#################
+@pytest.mark.dependency(name="test_checksums")
+@pytest.mark.parametrize(
+    "checksum_file, checksum, expectation",
+    [
+        pytest.param(
+            Path("md5checksums.txt"), "40df91d5c40cb55621c4c92201da6834", does_not_raise(), id="Normal case"
+        ),
+        pytest.param(
+            Path("malformed_md5_checksums.txt"),
+            None,
+            pytest.raises(ValueError),
+            id="Bad md5 file - tab separated",
+        ),
+        pytest.param(Path("dir_name"), None, does_not_raise(), id="Dir passed instead of file"),
+    ],
+)
+def test_checksums(
+    data_dir: Path, checksum_file: Optional[Path], checksum: str, expectation: ContextManager
```

Comment:
```suggestion
    data_dir: Path, checksum_file: Path, checksum: Optional[str], expectation: ContextManager
```
I believe I messed up the `Optional` here, sorry! (highlighted by `mypy`)